dev/ProgressRing/ProgressRing.cpp

```diff
@@ -0,0 +1,119 @@
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+#include "common.h"
+#include "ProgressRing.h"
+#include "RuntimeProfiler.h"
+#include "ResourceAccessor.h"
+#include "math.h"
+
+ProgressRing::ProgressRing()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_ProgressRing);
+
+    SetDefaultStyleKey(this);
+
+    RegisterPropertyChangedCallback(winrt::RangeBase::ValueProperty(), { this, &ProgressRing::OnRangeBasePropertyChanged });
+
+    SizeChanged({ this, &ProgressRing::OnSizeChanged });
+}
+
+void ProgressRing::OnApplyTemplate()
+{
+    winrt::IControlProtected controlProtected{ *this };
+
+    m_layoutRoot.set(GetTemplateChildT<winrt::Grid>(s_LayoutRootName, controlProtected));
+    m_outlineFigure.set(GetTemplateChildT<winrt::Windows::UI::Xaml::Media::PathFigure>(s_OutlineFigureName, controlProtected));
+    m_outlineArc.set(GetTemplateChildT<winrt::Windows::UI::Xaml::Media::ArcSegment>(s_OutlineArcName, controlProtected));
+    m_barFigure.set(GetTemplateChildT<winrt::Windows::UI::Xaml::Media::PathFigure>(s_BarFigureName, controlProtected));
+    m_barArc.set(GetTemplateChildT<winrt::Windows::UI::Xaml::Media::ArcSegment>(s_BarArcName, controlProtected));
+
+    RenderAll();
+}
+
+void  ProgressRing::OnPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    winrt::IDependencyProperty property = args.Property();
+    
+    // TODO: Implement
+}
+
+void ProgressRing::OnSizeChanged(const winrt::IInspectable&, const winrt::IInspectable&)
+{
+    RenderAll();
+}
+
+void ProgressRing::OnRangeBasePropertyChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args)
+{
+    RenderSegment();
+}
+
+void ProgressRing::OnStrokeThicknessPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    RenderAll();
+}
+
+winrt::Windows::Foundation::Size ProgressRing::ComputeEllipseSize(const double thickness)
+{
+    const double safeThickness = std::max(thickness, static_cast<double>(0.0));
+    const double actualWidth = ProgressRing::Width();
+    const double width = std::max((ProgressRing::Width() - safeThickness) / 2.0, 0.0);
+    const double height = std::max((ProgressRing::Height() - safeThickness) / 2.0, 0.0);
+
+    return winrt::Windows::Foundation::Size(static_cast<float>(width), static_cast<float>(height));
+}
+
+
+void ProgressRing::RenderSegment()
+{
+    if (auto&& progressRing = m_layoutRoot.get())
+    {
+        auto&& barFigure = m_barFigure.get();
+        auto&& barArc = m_barArc.get();
+        const double thickness = ProgressRing::StrokeThickness();
+        const double maximum = Maximum();
+        const double minimum = Minimum();
+
+        const double range = maximum - minimum;
+        const double delta = Value() - minimum;
+
+        double normalizedRange = (range == 0.0) ? 0.0 : (delta / range);
+        normalizedRange = std::min(std::max(0.0, normalizedRange), 0.9999);
```

Comment:
>0.9999 [](start = 67, length = 6)

what is this?

Reply:
I referenced what was in the community toolkit for the calculations.

Reply:
@StephenLPeters Turns out line 82 offsets the normalized range to display a full ring at 100%. It displays an empty ring if it is removed, or if I set the second value to 1.0 instead of 0.9999.

Reply:
Interesting! can you add a comment explaining this and perhaps use one of the standard library functions for finding next float values, like std::nextafter(1.0, 0.0);

Reply:
@StephenLPeters Turns out should `std::nextafter(1.0, 0.0)` is used, it can only be up to the precision of a float, or it still shows an empty ring at value = 100%